src/python.cpp

```diff
@@ -432,22 +465,129 @@ PyObject* py_import(const std::string& module) {
   return PyImport_Import(module_str);
 }
 
+
+class PyErrorScopeGuard {
+private:
+  PyObject *er_type, *er_value, *er_traceback;
+  bool pending_restore;
+
+public:
+  PyErrorScopeGuard() {
+    PyErr_Fetch(&er_type, &er_value, &er_traceback);
+    pending_restore = true;
+  }
+
+  void release(bool restore = false) {
+    if (restore)
+      PyErr_Restore(er_type, er_value, er_traceback);
+    pending_restore = false;
+  }
+
+  ~PyErrorScopeGuard() {
+    if (pending_restore)
+      PyErr_Restore(er_type, er_value, er_traceback);
+  }
+};
+// copied directly from purrr; used to call rlang::trace_back() in
+// py_fetch_error() in such a way that it doesn't introduce a new
+// frame in returned traceback
+SEXP current_env(void) {
+  static SEXP call = NULL;
+
+  if (!call) {
+    // `sys.frame(sys.nframe())` doesn't work because `sys.nframe()`
+    // returns the number of the frame in which evaluation occurs. It
+    // doesn't return the number of frames on the stack. So we'd need
+    // to evaluate it in the last frame on the stack which is what we
+    // are looking for to begin with. We use instead this workaround:
+    // Call `sys.frame()` from a closure to push a new frame on the
+    // stack, and use negative indexing to get the previous frame.
+    ParseStatus status;
+    SEXP code = PROTECT(Rf_mkString("sys.frame(-1)"));
+    SEXP parsed = PROTECT(R_ParseVector(code, -1, &status, R_NilValue));
+    SEXP body = VECTOR_ELT(parsed, 0);
+
+    SEXP fn = PROTECT(Rf_allocSExp(CLOSXP));
+    SET_FORMALS(fn, R_NilValue);
+    SET_BODY(fn, body);
+    SET_CLOENV(fn, R_BaseEnv);
+
+    call = Rf_lang1(fn);
+    R_PreserveObject(call);
+
+    UNPROTECT(3);
+  }
+
+  // Rf_PrintValue(get_r_trace(false, false));
+
+  return Rf_eval(call, R_BaseEnv);
+}
+
+SEXP eval_call(SEXP r_func, SEXP arg) {
+  SEXP cl = Rf_lang2(r_func, arg);
+  RObject cl_(cl); // protect
+  return Rf_eval(cl, ns_reticulate);
+}
+
+SEXP eval_call(SEXP r_func, SEXP arg1, SEXP arg2) {
+  SEXP cl = Rf_lang3(r_func, arg1, arg2);
+  RObject cl_(cl); // protect
+  return Rf_eval(cl, ns_reticulate);
+}
+
+SEXP eval_call(SEXP r_func, SEXP arg1, bool arg2) {
+  return eval_call(r_func, arg1, Rf_ScalarLogical(arg2));
+}
+
+SEXP eval_call_in_userenv(SEXP r_func, SEXP arg) {
+  SEXP cl = Rf_lang2(r_func, arg);
+  RObject cl_(cl); // protect
+  return Rf_eval(cl, current_env());
```

Comment:
`current_env()` should always return something that's already protected. (because it's on the R call stack)
It also isn't quite right for enabling `UseMethod("r_to_py")` to find methods in the calling env, since it resolves a reticualte internal frame instead, we'll need a new function for that anyway.